### PR TITLE
[modbus] BaseModbusThingHandler: Add ability to retrieve slave address

### PR DESCRIPTION
**Pull request by fwolter**
I think all features of ModbusEndpointThingHandler are now covered by BaseModbusThingHandler. That is why I made the getter to the ModbusEndpointThingHandler private again. Do you agree @ssalonen?

Signed-off-by: Fabian Wolter <github@fabian-wolter.de>

### CODE REVIEW

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:59]: Hmm this will now raise unchecked exception when bridge is offline e.g due to configuration error? When bridge is having configuration error, the slave id is not available.

Perhaps the configuration error of bridge is the only situation when these methods can fail?

Should it set thing to offline with BRIDGE_OFFLINE instead of raising unchecked exception to framework?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:184]: Consider renaming the method as it no longer returns endpoint

**Comment by ssalonen** [comment]
Makes sense to have it private, I agree.

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:59]: I see no other chance to prevent the Thing from being initialized. Setting the Thing status has no effect as it will be overwritten by any `updateStatus()` in `initialize()` of the "user" class. Do you have a suggestion?

Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:59]: Good point... 

I looked at some examples from the code base

- `AbstractModbusEndpointThingHandler` parent class sets thing OFFLINE. initialize not implemented in child class
- `BeaconBluetoothHandler` (in bluetooth) parent class sets thing OFFLINE. Child class (RuuviTagHandler) has check 
```
        if (getThing().getStatus() != ThingStatus.OFFLINE) {
            heartbeatFuture = scheduler.scheduleWithFixedDelay(this::heartbeat, 0, HEARTBEAT_TIMEOUT_MINUTES,
                    TimeUnit.MINUTES);
```
I there might be race condition here... Ist getThing().getStatus() update instantaneous?
- `IT100BridgeHandler extends DSCAlarmBaseBridgeHandler` super initialize is called conditionally when config is otherwise OK.

Do we typically expect to override the initialize method in child class?

Comment (fwolter) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:59]: It's always good to have a second pair of eyes. The examples inspired me to come up with a slightly other solution.

The BaseModbusThingHandler got a new abstract method `modbusInitialize()`, which is only called, when the Modbus Bridge is configured correctly. That way the Thing can be set to offline in a normal way. Also, the user doesn't need to call `super.initialize()` anymore, which was error-prone. So, the boolean flag could be dropped, too. I made the `initialize()` method `final` to prevent the user from overriding it wrongly.

**Comment by cpmeister** [comment]
Is this PR ready to merge?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:75]: I don't think this is true?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:66]: BRIDGE_OFFLINE would be more correct for cases when bridge is offline? It would direct the user to look at the bridge error message.

Are there any other reasons we expect the initialization to fail?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (ssalonen) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:184]: According to javadoc, null communication interface is returned when initialization is incomplete == bridge is not online.

By failing fast (check bridge status) we can offer better error message to user?

**Review by ssalonen — COMMENTED** [review]
Checked the code again, good change with the final initialize.

I think the thing status and error handling could use little work still, what do you think?

**Comment by fwolter** [comment]
From my side it's ready to merge. @ssalonen are your comments addressed?

**Comment by ssalonen** [comment]
@fwolter i had some minor comments on error messages still to guide the user. What do you think?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:75]: You're right. I removed it.

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:66]: I saw there's a ThingStatusDetail.BRIDGE_UNINITIALIZED. I think that would fit even better. I moved the invocation of the abstract method out of the try-block, that only a Bridge error could lead to a BRIDGE_UNINITIALIZED.

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/BaseModbusThingHandler.java:184]: I reworked all error message to achieve easier debugging for the user.

**Review by ssalonen — APPROVED** [review]
You are fast! Lgtm.

**Review by cpmeister — APPROVED** [review]

**Comment by cpmeister** [comment]
I want the modbus integration tests to succeed before I merge this.

**Pull request merged**
